ext/pdo_sqlite/pdo_sqlite.c

```diff
@@ -21,39 +21,654 @@
+	size_t func_name_len;
+	zend_long argc = -1;
+	zend_long flags = 0;
+	pdo_dbh_t *dbh;
+	pdo_sqlite_db_handle *H;
+	int ret;
+
+	ZEND_PARSE_PARAMETERS_START(2, 4)
+			Z_PARAM_STRING(func_name, func_name_len)
+			Z_PARAM_FUNC(fci, fcc)
+			Z_PARAM_OPTIONAL
+			Z_PARAM_LONG(argc)
+			Z_PARAM_LONG(flags)
+	ZEND_PARSE_PARAMETERS_END();
+
+	dbh = Z_PDO_DBH_P(ZEND_THIS);
+	PDO_CONSTRUCT_CHECK;
+
+	H = (pdo_sqlite_db_handle *)dbh->driver_data;
+	func = (struct pdo_sqlite_func*)ecalloc(1, sizeof(*func));
+
+	ret = sqlite3_create_function(H->db, func_name, argc, flags | SQLITE_UTF8,
+								  func, php_pgsql_func_callback, NULL, NULL);
+	if (ret == SQLITE_OK) {
+		func->funcname = estrdup(func_name);
+
+		ZVAL_COPY(&func->func, &fci.function_name);
+
+		func->argc = argc;
+
+		func->next = H->funcs;
+		H->funcs = func;
+
+		RETURN_TRUE;
+	}
+
+	efree(func);
+	RETURN_FALSE;
+}
+
+#ifndef PDO_SQLITE_OMIT_LOAD_EXTENSION
+/* Attempts to load an SQLite extension library. */
+PHP_METHOD(PdoSqlite, loadExtension)
+{
+	char *extension, *errtext = NULL;
+	char fullpath[MAXPATHLEN];
+	size_t extension_len;
+
+	pdo_dbh_t *dbh;
+	pdo_sqlite_db_handle *db_handle;
+
+	if (zend_parse_parameters(ZEND_NUM_ARGS(), "s", &extension, &extension_len) == FAILURE) {
+		RETURN_THROWS();
+	}
+
+	if (extension_len == 0) {
+		zend_argument_value_error(1, "cannot be empty");
+		RETURN_THROWS();
+	}
+
+	dbh = Z_PDO_DBH_P(ZEND_THIS);
+	PDO_CONSTRUCT_CHECK;
+
+	db_handle = (pdo_sqlite_db_handle *)dbh->driver_data;
+
+#ifdef ZTS
+	if ((strncmp(sapi_module.name, "cgi", 3) != 0) &&
+		(strcmp(sapi_module.name, "cli") != 0) &&
+		(strncmp(sapi_module.name, "embed", 5) != 0)
+	) {
+		// TODO - needs test.
+		zend_throw_exception_ex(php_pdo_get_exception(), 0, "Not supported in multithreaded Web servers");
+		RETURN_THROWS();
+	}
+#endif
+
+	if (!VCWD_REALPATH(extension, fullpath)) {
+		zend_throw_exception_ex(php_pdo_get_exception(), 0, "Unable to load extension \"%s\"", extension);
+		RETURN_THROWS();
+	}
+
+	sqlite3 *sqlite_handle;
+	sqlite_handle = db_handle->db;
+
+	// This only enables extension loading for the C api, not for SQL
+	sqlite3_db_config(sqlite_handle, SQLITE_DBCONFIG_ENABLE_LOAD_EXTENSION, 1, NULL);
+
+	if (sqlite3_load_extension(sqlite_handle, fullpath, 0, &errtext) != SQLITE_OK) {
+		zend_throw_exception_ex(php_pdo_get_exception(), 0, "Unable to load extension \"%s\"", errtext);
+		sqlite3_free(errtext);
+		sqlite3_db_config(sqlite_handle, SQLITE_DBCONFIG_ENABLE_LOAD_EXTENSION, 0, NULL);
+		RETURN_THROWS();
+	}
+
+	// We disable extension loading for a vague feeling of safety. This is probably not necessary
+	// as extensions can only be loaded through C code, not through SQL, and if someone can get
+	// some C code to run on the server, they can do anything.
+	sqlite3_db_config(sqlite_handle, SQLITE_DBCONFIG_ENABLE_LOAD_EXTENSION, 0, NULL);
+}
+#endif
+
+typedef struct {
+	sqlite3_blob *blob;
+	size_t		 position;
+	size_t       size;
+	int          flags;
```

Comment:
Should probably unsigned?

Reply:
It copy pastes the similar struct from ext/sqlite3, and I don't know if it's a good idea to alter from it. But apart from this, it looks like that it can be unsigned,

Reply:
Although I'm wondering why not to export and the reuse the original struct? I guess this would be the best idea... 🤔

Reply:
Right, well maybe it also makes sense to have it unsigned in ext/sqlite3, while ``^``, ``~``, ``&`` and ``|`` are well defined even on signed, ``<<`` is UB and ``>>`` is implementation defined sooo

Reply:
Fine, since the struct is not exposed and the constant values (like `SQLITE_OPEN_EXRESCODE`) won't overflow, I'm fine with changing the type for ext/sqlite3 as well

Reply:
This can wait for a follow-up PR in all honesty :)

Reply:
FYI
It looks okay with unsigned.
https://github.com/sqlite/sqlite/blob/95cf95841c81a777a9268a1a375b30b6a8466628/src/sqlite.h.in#L593

Reply:
I thought there would be no problem, but sqlite3 internally manages flags as ints, so it might be safer to use the same specifications.
https://github.com/sqlite/sqlite/blob/95cf95841c81a777a9268a1a375b30b6a8466628/src/os_unix.c#L6242

Reply:
ACK